### PR TITLE
[RNMobile] Gallery - Tiles component

### PR DESCRIPTION
**Pull request by mkevins**
## Description

This PR introduces a Tiles component for the semi-cross-platform Gallery block. The purpose of this component is to introduce a responsive layout abstraction for mobile views.

### PR Hierarchy
This PR is part of the [PR hierarchy described here](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1416#issuecomment-551057738). This PR can be tested with the aggregate changeset and integration of components within the related PRs by checking out the branch of the "top level" PR: https://github.com/WordPress/gutenberg/pull/18265.

## To test
Test this component by checking out the branch of the "top level" PR: https://github.com/WordPress/gutenberg/pull/18265.

Open the gallery settings from the demo app and change the "Columns" value (in the BottomSheet) to different values.

#### Expected behavior in portrait

* Tiles layout should update to reflect the columns setting (visible columns should not exceed 2, but will go down to 1)
* Tiles should be evenly spaced
* Tiles in last row should expand to fill available width

#### Expected behavior in landscape

* Tiles layout should update to reflect the columns setting
* Tiles should be evenly spaced
* Tiles in last row should expand to fill available width

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/block-library/src/gallery/tiles.native.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	flexDirection: 'row',
+	alignItems: 'center',
+	borderColor: 'transparent',
+};
+
+function Tiles( props ) {
+	const {
+		columns,
+		children,
+		groutSpacing = 10,
```
Comment (koke) [inline packages/block-library/src/gallery/tiles.native.js]: This one looked like a typo until I decided to check the dictionary first (TIL 😁). I think `spacing` works as well and might be clearer/shorter.

File: packages/block-library/src/gallery/tiles.native.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		const rowLength = row === lastRow ? ( lastTile % columns ) + 1 : columns;
+		const indexInRow = index % columns;
+
+		return (
+			<View style={ [ tileStyle, {
+				width: ( ( 1 / rowLength ) * 100 ) + '%',
+				borderLeftWidth: groutSpacing * ( indexInRow / rowLength ),
+				borderRightWidth: groutSpacing * ( 1 - ( ( indexInRow + 1 ) / rowLength ) ),
+				borderTopWidth: row === 0 ? 0 : groutSpacing / 2,
+				borderBottomWidth: row === lastRow ? 0 : groutSpacing / 2,
```
Comment (koke) [inline packages/block-library/src/gallery/tiles.native.js]: Why are we setting these as borders and not margins?

File: packages/block-library/src/gallery/tiles.native.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	flexDirection: 'row',
+	flexWrap: 'wrap',
+};
+
+const tileStyle = {
+	overflow: 'hidden',
+	flexDirection: 'row',
+	alignItems: 'center',
+	borderColor: 'transparent',
+};
```
Comment (koke) [inline packages/block-library/src/gallery/tiles.native.js]: These work, but if we're not going to use scss files, we should at least use `StyleSheet.create` to create these objects, as it adds some extra validation and type hinting.

**Review by koke — COMMENTED** [review]
This is looking pretty solid. I found some discrepancies on image sizing, but I think those are more related to `GalleryImage` and not the `Tiles` component.

The logic to distribute the spaces evenly was not obvious from the start. I had to run it through the debugger and simulate a few scenarios before I realized what was going on. It is also different from what the web does. I went to suggest simpler alternatives, but then I realized that without a native alternative to `calc()`, this was our best way to ensure equal widths and spacing for all images.
So the implementation is good, but still hard to follow, and I think it could benefit from some comments explaining what it's trying to do and why.

Interestingly enough, I tested in RTL mode and react-native seems to automatically flip those border widths correctly 😮

File: packages/block-library/src/gallery/tiles.native.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	const wrappedChildren = Children.map( children, ( child, index ) => {
+		const row = Math.floor( index / columns );
+		const rowLength = row === lastRow ? ( lastTile % columns ) + 1 : columns;
+		const indexInRow = index % columns;
+
+		return (
+			<View style={ [ tileStyle, {
+				width: ( ( 1 / rowLength ) * 100 ) + '%',
+				borderLeftWidth: groutSpacing * ( indexInRow / rowLength ),
+				borderRightWidth: groutSpacing * ( 1 - ( ( indexInRow + 1 ) / rowLength ) ),
```
Comment (koke) [inline packages/block-library/src/gallery/tiles.native.js]: Comparing indexes and lengths always gets annoying 😁 In this case, if you defined `lastIndexInRow = rowLength - 1` and use that as the denominator in both, you can skip the index offset and the code looks clearer IMO: the right border would read more clearly as `groutSpacing - borderLeft`.

File: packages/block-library/src/gallery/tiles.native.js (hunk 1 above)
Comment (mkevins) [reply packages/block-library/src/gallery/tiles.native.js]: Perhaps I took the metaphor a bit too far :laughing:. I changed this to `spacing`.

File: packages/block-library/src/gallery/tiles.native.js (hunk 2 above)
Comment (mkevins) [reply packages/block-library/src/gallery/tiles.native.js]: I used borders for this so that fixed sizes would be included in the total width of each item, preserving the correct wrapping behavior. I've added a note about this in the code comment.

File: packages/block-library/src/gallery/tiles.native.js (hunk 3 above)
Comment (mkevins) [reply packages/block-library/src/gallery/tiles.native.js]: Using `rowLength - 1` as a denominator can be problematic for rows that only have one tile (divide by zero). The underlying problem I've struggled with in simplifying this further is that the left and right spacings for a given tile are not linearly dependent: i.e. right border cannot be expressed solely as a function of the spacing and the left border. Instead, the right border is complementary with the left border of the _next_ tile (hence the fencepost calculations). I've added an explanation in the code comment.

File: packages/block-library/src/gallery/tiles.native.js (hunk 4 above)
Comment (mkevins) [reply packages/block-library/src/gallery/tiles.native.js]: Using `rowLength - 1` as a denominator can be problematic for rows that only have one tile (divide by zero). The underlying problem I've struggled with in simplifying this further is that the left and right spacings for a given tile are not linearly dependent: i.e. right border cannot be expressed solely as a function of the spacing and the left border. Instead, the right border is complementary with the left border of the _next_ tile (hence the fencepost calculations). I've added an explanation in the code comment.

**Comment by mkevins** [comment]
> The logic to distribute the spaces evenly was not obvious from the start. I had to run it through the debugger and simulate a few scenarios before I realized what was going on. It is also different from what the web does. I went to suggest simpler alternatives, but then I realized that without a native alternative to `calc()`, this was our best way to ensure equal widths and spacing for all images.
> So the implementation is good, but still hard to follow, and I think it could benefit from some comments explaining what it's trying to do and why.

I agree, and I've added an explanatory code comment for the layout logic used here.

**Review by koke — APPROVED** [review]
Thanks for the great explanation :shipit:

File: packages/block-library/src/gallery/tiles.native.js (hunk 2 above)
Comment (koke) [reply packages/block-library/src/gallery/tiles.native.js]: I didn't realize I had said margins here when I finished the review. Why not paddings? I would think they offer the same effect but they would be more meaningful than transparent borders

Comment (mkevins) [reply packages/block-library/src/gallery/tiles.native.js]: I think padding is semantically better than border for what we are doing here, and I even used padding in an earlier iteration (from what I can tell it works pretty much the same way). The reason I switched back to using border is that I had the thought that border would be less likely to be used later if we wanted to modify the styles, so it was more "disposable" in that sense (i.e. by not using padding, we still have the opportunity to use it for its original purpose, if that need arises).

**Comment by mkevins** [comment]
Thanks for reviewing Koke!

FYI: Just added an empty commit to trigger travis to re-run tests (the re-run button seemingly isn't working).

File: packages/block-library/src/gallery/tiles.native.js (hunk 2 above)
Comment (pinarol) [reply packages/block-library/src/gallery/tiles.native.js]: > (i.e. by not using padding, we still have the opportunity to use it for its original purpose, if that need arises).

Could you help me understand this better? What kind of need for example? Using borders for spacing purposes is not very intuitive for other people who will maintain it. And having the need to put such a big code comment can be signaling that we are introducing a "hard to understand" code. So I'd prefer using paddings unless we have good reasons.

Comment (pinarol) [reply packages/block-library/src/gallery/tiles.native.js]: This looks simpler on web, only right and bottom margins are used to give the spacing and the bottom margin is a constant. Only thing that is varies is the right margin, it is just set to 0 for the right most tile.

galleryTiles2

Can we use a similar approach?

File: packages/block-library/src/gallery/tiles.native.js (hunk 1 above)
Comment (pinarol) [reply packages/block-library/src/gallery/tiles.native.js]: I agree that spacing work better. I should say though, when I found out what grout means I had a lol moment, I appreciate how creative this metaphor is :D

File: packages/block-library/src/gallery/tiles.native.js (hunk 2 above)
Comment (mkevins) [reply packages/block-library/src/gallery/tiles.native.js]: Hi @pinarol :wave: thanks for taking a look at this PR.

> Could you help me understand this better? What kind of need for example?

We are currently using borders to evenly distribute space between tiles. If we want (in the future) to give each tile padding, we can, since that property is not being used for this layout purpose.

> And having the need to put such a big code comment can be signaling that we are introducing a "hard to understand" code. So I'd prefer using paddings or margins unless we have good reasons.

FWIW, the use of `padding` instead of `border` won't alleviate the need for the comment. AFAIK, the use of margins is not possible for this. The aim of the comment is to make it _easier_ to understand for future maintainers.

> This looks simpler on web, only right and bottom margins are used to give the spacing and the bottom margin is a constant. Only thing that varies is the right margin, it is just set to 0 for the right most tile.

Right, web is able to use margins. We are not able to do this, since margins won't be included in the width calculations (breaking the wrapping behavior).

Comment (pinarol) [reply packages/block-library/src/gallery/tiles.native.js]: > Right, web is able to use margins. We are not able to do this, since margins won't be included in the width calculations (breaking the wrapping behavior).

I see but It is possible if we change the approach a bit, it just requires us to calculate the tile width in pixels rather than percentages. 

Actually I think the most challenging part of this manual calculation is finding the `rowLength`, and you already did it for "calculating the width based on percentage" solution.

I still think we can make the current code easier to understand and maintain.

I'll drop my example here too, for documenting purposes. Let's give this a try and see how it goes. I added a round down in 0,00 precision so I think that'll help with exceeding the container width.

```

class Tiles extends Component {

	constructor() {
		super( ...arguments );
		this.onLayout = this.onLayout.bind( this );

		this.state = {
			containerWidth: 0,
		};
	}
		
	onLayout( event ) {
		const { width, height } = event.nativeEvent.layout;
		this.setState( {
			containerWidth: width,
		} );
	}

	render() {
		const {
			columns,
			children,
			spacing = 10,
		} = this.props;
		
		const tileCount = Children.count( children );
		const lastTile = tileCount - 1;
		const lastRow = Math.floor( lastTile / columns );

		const wrappedChildren = Children.map( children, ( child, index ) => {
			const { containerWidth } = this.state;
			if ( containerWidth === 0 ) {
				return null;
			}

			const rowLength = (() => {
				const row = Math.floor( index / columns );
				const rowLength = row === lastRow ? ( lastTile % columns ) + 1 : columns;	
				return rowLength;
			})();

			const tileWidth = (() => {
				const totalHorizontalMargin = ( rowLength - 1 ) * spacing;
				const width =  ( containerWidth - totalHorizontalMargin ) / rowLength;
				return Math.floor( width * 100 ) / 100; // round down to 2 decimals
			})();

			const isRightMostTile = ( index === lastTile ) || (( index + 1 ) % rowLength === 0);

			return (
				<View style={ [ styles.tileStyle, {
					width: tileWidth,
					marginRight: isRightMostTile ? 0 : spacing,
					marginBottom: spacing,
				} ] }>
					{ child }
				</View>
			);
		} );
	
		return (
			<View style={ styles.containerStyle } onLayout={ this.onLayout }>
				{ wrappedChildren }
			</View>
		);
	}
}

```

Comment (mkevins) [reply packages/block-library/src/gallery/tiles.native.js]: > Let's give this a try and see how it goes.

I tried this out, but I'm still seeing problems on orientation changes, and columns (in landscape).

**Screenshots:**

Rotation:


Columns in landscape:

Comment (pinarol) [reply packages/block-library/src/gallery/tiles.native.js]: It looks like I made a mistake on this line:

`const isRightMostTile = ( index === lastTile ) || (( index + 1 ) % rowLength === 0);`

The correct one should be:

`const isRightMostTile = ( index === lastTile ) || (( index + 1 ) % columns === 0);`

It is working fine with this fix according to my experiments.

BTW I am fine with both:
- going with percentages and using paddings
 or 
- going with margins and calculating the width in pixels

I just wouldn't prefer to go with transparent borders as it is not what they are for and for that reason it is making code harder to understand. And we can always iterate if a special need comes up that will require us to use borders for spacing.

When it comes to paddings vs margins, they have both pros/cons. Advantage of padding solution is we don't need the onLayout handling. 

Advantage of margin solution is, it is making the space calculation very easy, the only thing to calculate is left as: `marginRight: isRightMostTile ? 0 : spacing,`. 

Personally I'd go with margins but I am fine with both solutions. 👍 

I wouldn't expect any performance differences between both ways because whenever container width changes the tiles should be re-rendered whether they are calculated in pixels or percentages.

File: packages/block-library/src/gallery/tiles.native.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+		 * [tile n + 1] will be equal for all tiles except the last one in a given
+		 * row).
+		 */
+
+		const row = Math.floor( index / columns );
+		const rowLength = row === lastRow ? ( lastTile % columns ) + 1 : columns;
+		const indexInRow = index % columns;
+
+		return (
+			<View style={ [ styles.tileStyle, {
```
Comment (pinarol) [inline packages/block-library/src/gallery/tiles.native.js]: We should set a key here otherwise we are getting a severe warning:

Screen Shot 2019-11-25 at 19 44 22

Adding `key={ index }` should solve it.

File: packages/block-library/src/gallery/tiles.native.js (hunk 2 above)
Comment (mkevins) [reply packages/block-library/src/gallery/tiles.native.js]: Hi @pinarol :wave: 

> The correct one should be:
> 
> `const isRightMostTile = ( index === lastTile ) || (( index + 1 ) % columns === 0);`
> 

I've tried the margin approach again with these suggested changes, however, I'm still seeing the same layout problem on orientation changes [described previously](https://github.com/WordPress/gutenberg/pull/18130#discussion_r345010804).

> BTW I am fine with both:
> 
> * going with percentages and using paddings
>   or
> * going with margins and calculating the width in pixels
> 
> I just wouldn't prefer to go with transparent borders as it is not what they are for and for that reason it is making code harder to understand. And we can always iterate if a special need comes up that will require us to use borders for spacing.
> 

Unfortunately, neither border nor padding are semantically correct in how they're being used here, according to the box model. I don't have a strong opinion about which is "better" between `border` and `padding`, so I've changed it to padding.

> When it comes to paddings vs margins, they have both pros/cons. Advantage of padding solution is we don't need the onLayout handling.
> 
> Advantage of margin solution is, it is making the space calculation very easy, the only thing to calculate is left as: `marginRight: isRightMostTile ? 0 : spacing,`.
> 
> Personally I'd go with margins but I am fine with both solutions. 
> 
> I wouldn't expect any performance differences between both ways because whenever container width changes the tiles should be re-rendered whether they are calculated in pixels or percentages.

Another advantage to the current spacing interpolation approach is that it allows `Tiles` to remain a stateless functional component instead of a stateful class component.

Since the `onLayout` approach seems to result in an unstable layout during orientation changes, I've gone with the interpolation approach. I've pushed a change to use `padding` instead of `border`, and updated the comment to reflect that.

File: packages/block-library/src/gallery/tiles.native.js (hunk 5 above)
Comment (mkevins) [reply packages/block-library/src/gallery/tiles.native.js]: Thanks for documenting that. Hm.. :thinking: that's unexpected.

> Adding key={ index } should solve it.

The weird thing is that this should be the default. From [the docs](https://reactjs.org/docs/lists-and-keys.html#keys):

> If you choose not to assign an explicit key to list items then React will default to using indexes as keys.

I haven't been able to reproduce it. Can you provide steps to reproduce it?

File: packages/block-library/src/gallery/tiles.native.js (hunk 2 above)
Comment (pinarol) [reply packages/block-library/src/gallery/tiles.native.js]: > I've tried the margin approach again with these suggested changes, however, I'm still seeing the same layout problem on orientation changes described previously.

I am super curious why that's happening now because I was able to repro the problem and that change seemed fixing that. If you want we can work on the problem together. But as I said before I am also fine with using paddings. 

> Unfortunately, neither border nor padding are semantically correct in how they're being used here

You have a point, but I wouldn't consider paddings and borders as _equally_ incorrect if they are used for spacing purposes. And if we are looking for semantically the most correct one that seems to be margins.

> Since the onLayout approach seems to result in an unstable layout during orientation changes, I've gone with the interpolation approach. 

If you mean [this](https://github.com/WordPress/gutenberg/pull/18130#discussion_r345010804) problem by unstable layout I think that's not because of onLayout, that one is just a result of an immature solution(I spent only ~30mins) and the logic that calculates the right most tile can have flaws, it just needs some tackling. Or is there any other problem that I am missing?

File: packages/block-library/src/gallery/tiles.native.js (hunk 5 above)
Comment (pinarol) [reply packages/block-library/src/gallery/tiles.native.js]: I am trying with Android emulator Pixel 2XL and enabled Remote JS Debugging.

I created a Gallery with 7 images.

Whenever I open the post or rotate the device that error is printed in Chrome dev tools console.

Thanks for the docs btw. We can use web's approach to set keys, it is a lot better as far as I see: `key={ img.id || img.url }` (from .../src/gallery/gallery.js).

Comment (mkevins) [reply packages/block-library/src/gallery/tiles.native.js]: I tried with those steps, but was still unable to reproduce this.

> We can use web's approach to set keys, it is a lot better as far as I see: key={ img.id || img.url }

We are currently using [this logic as well in gallery.native.js](https://github.com/WordPress/gutenberg/pull/18176/files#diff-c224ac948fbbffd7b92bbb5e04763319R66) (though there may be [some concerns about it](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1610#issuecomment-558532685)). Since we are currently wrapping the children in a `View`, I'd expect the children to be assigned their default key as index. Do you have the html content of the post where you see this warning? Perhaps the content can provide a clue about how this is occurring, or make it possible to reproduce.

It's not clear why React is not using the default `index` as key, but another possibility is to elevate the key to the child's wrapping view:

```diff
diff --git a/packages/block-library/src/gallery/tiles.native.js b/packages/block-library/src/gallery/tiles.native.js
index f1666db..895f126 100644
--- a/packages/block-library/src/gallery/tiles.native.js
+++ b/packages/block-library/src/gallery/tiles.native.js
@@ -56,7 +56,7 @@ function Tiles( props ) {
 		const indexInRow = index % columns;
 
 		return (
-			<View style={ [ styles.tileStyle, {
+			<View key={ child.key } style={ [ styles.tileStyle, {
 				width: `${ 100 / rowLength }%`,
 				paddingLeft: spacing * ( indexInRow / rowLength ),
 				paddingRight: spacing * ( 1 - ( ( indexInRow + 1 ) / rowLength ) ),
```

But, since I can't reproduce the original warning, I couldn't confirm whether this addresses the issue you've observed.

File: packages/block-library/src/gallery/tiles.native.js (hunk 2 above)
Comment (mkevins) [reply packages/block-library/src/gallery/tiles.native.js]: > Or is there any other problem that I am missing?

I'm not sure. fwiw, your logic to determine the right-most tile looks correct to me. In fact, I had earlier implementations that used a similar approach with `margin`, and even `flex-basis`, but all of them suffered from layout problems on orientation changes. The only solution that seems robust in my tests is the current interpolation approach. When using the `onLayout` approach, adding `console.log(width)` showed some differences in the layout calculations:

```
Starting app in landscape:

2x 358.5454406738281 tiles.native.js:35
5x 360.7272644042969 tiles.native.js:35

After orientation change:

3x 548               tiles.native.js:35
3x 546.1818237304688 tiles.native.js:35
```

For each, `onLayout` is called 7 and 6 times, respectively, for portrait and landscape orientations. It seems that the width "adjusts" as the calculation "settles" on a final result. In the case of portrait, width seems to grow slightly, whereas in landscape it seems to shrink.

File: packages/block-library/src/gallery/tiles.native.js (hunk 5 above)
Comment (pinarol) [reply packages/block-library/src/gallery/tiles.native.js]:
```
<!-- wp:gallery {"ids":["708","415","701","728","684","682","416",555,"737"],"columns":3,"linkTo":"media"} -->
<figure class="wp-block-gallery columns-3 is-cropped"><ul class="blocks-gallery-grid"><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0111-9.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0111-9.jpg" data-id="708" class="wp-image-708"/></a><figcaption class="blocks-gallery-item__caption"><p>Caption</p></figcaption></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/09/1487352383_mu-shsjvgty-4-1.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/09/1487352383_mu-shsjvgty-4-1.jpg" data-id="415" class="wp-image-415"/></a><figcaption class="blocks-gallery-item__caption">Alşksdf şlkas dolaş dfasd şad fsa<br>Asf fasd fas<br>A sdfasdf sadf<br> Asdf</figcaption></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0003-5.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0003-5.jpg" data-id="701" class="wp-image-701"/></a><figcaption class="blocks-gallery-item__caption">Hello<br>World</figcaption></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0005-6.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0005-6.jpg" data-id="728" class="wp-image-728"/></a><figcaption class="blocks-gallery-item__caption">Asdf sadf asdf fşilaks dfişkl asdAsdf asf asdf<br>Asdf</figcaption></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0002-1-2.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0002-1-2.jpg" data-id="684" class="wp-image-684"/></a><figcaption class="blocks-gallery-item__caption">Caption yazıyorum gözlerim kapalı asdf asdf asf asf asdf asdf asdf asdf sd Asdf asdf asf f<br>Asf asdf <br>As fasdf<br>Asf fasdf </figcaption></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0004-7.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/img_0004-7.jpg" data-id="682" class="wp-image-682"/></a><figcaption class="blocks-gallery-item__caption">Selale selale asdf hklajsdhf lkajsdf lkajshdfl kasjdfhlkajhdf lkajsdf lkasjdfhl aksdjfh lkasjdhf laksjdhf lkasjhdfl ksajdfha lksdfhklasdjh ;Hal;skfjla;skinflint ;lkajsdf ;lkajsdf ;laskdjf </figcaption></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/09/7fe66fb63510d8798a9a5505838cecf2-2.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/09/7fe66fb63510d8798a9a5505838cecf2-2.jpg" data-id="416" class="wp-image-416"/></a></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/a3022813-7005-47b9-92e6-1004d36e3803.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/10/a3022813-7005-47b9-92e6-1004d36e3803.jpg" data-id="555" class="wp-image-555"/></a></figure></li><li class="blocks-gallery-item"><figure><a href="https://pinarspremiumtestsite123443.files.wordpress.com/2019/11/img_0111.jpg"><img src="https://pinarspremiumtestsite123443.files.wordpress.com/2019/11/img_0111.jpg" data-id="737" class="wp-image-737"/></a></figure></li></ul></figure>
<!-- /wp:gallery -->

<!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph -->
```

File: packages/block-library/src/gallery/tiles.native.js (hunk 2 above)
Comment (pinarol) [reply packages/block-library/src/gallery/tiles.native.js]: I am all fine with going with padding solution 👍 onLayout can be called multiple times but we don't need to update the component state if container width stays same. And if container width does change, the tiles will need layout again independent from whether they are calculated using pixels or percentages. 

But if you observed some odd behavior in case of margins let's go with paddings. 

I am up for both: going as is or working together to see what's wrong with margin solution(because I couldn't repro with my local env).

File: packages/block-library/src/gallery/tiles.native.js (hunk 5 above)
Comment (pinarol) [reply packages/block-library/src/gallery/tiles.native.js]: Interestingly I can't repro this on iOS simulators or in my Android device(Huawei p20 lite). I can only repro this with my Pixel 2XL emulator. I also checked web side with dev tools and it is not reproducible there either. Considering `We are currently using this logic as well in gallery.native.js` it seems like we don't have to block for this.

**Review by pinarol — APPROVED** [review]
I think this is ready to go! 🎉
